core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1164,36 +1167,81 @@ private void deleteInternal(LockedInodePath inodePath, boolean replayed, long op
+            safeRecursiveUFSDeletes.put(descendantPath, descendant);
+          } else {
+            // Invalidate ancestor directories if not a mount point
+            AlluxioURI currentPath = descendantPath;
+            while (currentPath.getParent() != null && !mMountTable.isMountPoint(currentPath)
+                && safeRecursiveUFSDeletes.containsKey(currentPath.getParent())) {
+              safeRecursiveUFSDeletes.remove(currentPath.getParent());
+              currentPath = currentPath.getParent();
+            }
+          }
+        }
+      }
+
+      // Inodes to delete from tree after attempting to delete from UFS
+      List<Inode> inodesToDelete = new LinkedList<>();
+      // Inodes that are not safe for recursive deletes
+      Set<Long> unsafeInodes = new HashSet<>();
+      // Alluxio URIs which could not be deleted
+      List<String> failedUris = new LinkedList<>();
 
       TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath);
       // We go through each inode, removing it from its parent set and from mDelInodes. If it's a
       // file, we deal with the checkpoints and blocks as well.
       for (int i = delInodes.size() - 1; i >= 0; i--) {
-        Inode<?> delInode = delInodes.get(i);
-        // the path to delInode for getPath should already be locked.
-        AlluxioURI alluxioUriToDel = mInodeTree.getPath(delInode);
+        Pair<AlluxioURI, Inode<?>> delInodePair = delInodes.get(i);
+        AlluxioURI alluxioUriToDel = delInodePair.getFirst();
+        Inode delInode = delInodePair.getSecond();
         tempInodePath.setDescendant(delInode, alluxioUriToDel);
 
-        // TODO(jiri): What should the Alluxio behavior be when a UFS delete operation fails?
-        // Currently, it will result in an inconsistency between Alluxio and UFS.
-        if (!replayed && delInode.isPersisted()) {
+        boolean failedToDelete = false;
+        if (unsafeInodes.contains(delInode.getId())) {
+          failedToDelete = true;
+        }
+        if (!replayed && delInode.isPersisted() && !failedToDelete) {
           try {
             // If this is a mount point, we have deleted all the children and can unmount it
             // TODO(calvin): Add tests (ALLUXIO-1831)
             if (mMountTable.isMountPoint(alluxioUriToDel)) {
               unmountInternal(alluxioUriToDel);
-            } else {
+            } else if (!alluxioOnly) {
               // Delete the file in the under file system.
               MountTable.Resolution resolution = mMountTable.resolve(alluxioUriToDel);
               String ufsUri = resolution.getUri().toString();
               UnderFileSystem ufs = resolution.getUfs();
-              boolean failedToDelete = false;
-              if (!alluxioOnly) {
+              AlluxioURI parentURI = alluxioUriToDel.getParent();
+              // Check if parent is deleted recursively
+              if (!safeRecursiveUFSDeletes.containsKey(parentURI)) {
```

Comment:
@gpang You're absolutely correct. I'm refactoring this to store only "unsafe" instead of "safe" directories.

Time / Space Complexity tradeoff:
- O (#safe) space / O(1) time for lookup isSafe
- O(#unsafe) space / O(log(n)) time for lookup isSafe. This can be reduced to O(1) with a trie-like data structure but I'll leave that for future optimization